src/digraph.rs

```diff
@@ -2265,6 +2274,159 @@ impl PyDiGraph {
+            while let Some(node) = dfs.next(&self.graph) {
+                if nodes.contains(&node) {
+                    // we found a path back to `nodes`
+                    return false;
+                }
+            }
+
+            true
+        };
+
+        let mut indices_to_remove: HashSet<NodeIndex> =
+            to_replace.iter().map(|&n| NodeIndex::new(n)).collect();
+
+        if check_cycle.unwrap_or(self.check_cycle)
+            && !can_contract(&indices_to_remove)
+        {
+            return Err(DAGWouldCycle::new_err(
+                "Substitution would create cycle(s)",
+            ));
+        }
+
+        // Create new node.
+        let node_index = self.graph.add_node(obj);
+
+        // Sanitize new node index from user input.
+        indices_to_remove.remove(&node_index);
+
+        // Determine edges for new node.
+        let mut incoming_edges: Vec<_> = indices_to_remove
+            .iter()
+            .flat_map(|&i| self.graph.edges_directed(i, Direction::Incoming))
+            .filter_map(|edge| {
+                let pred = edge.source();
+                if !indices_to_remove.contains(&pred) {
+                    Some((edge.source(), edge.weight().clone_ref(py)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let mut outgoing_edges: Vec<_> = indices_to_remove
+            .iter()
+            .flat_map(|&i| self.graph.edges_directed(i, Direction::Outgoing))
+            .filter_map(|edge| {
+                let succ = edge.target();
+                if !indices_to_remove.contains(&succ) {
+                    Some((edge.target(), edge.weight().clone_ref(py)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // Remove nodes that will be replaced.
+        for index in indices_to_remove {
+            self.graph.remove_node(index);
+        }
+
+        // If `weight_combo_fn` was specified, merge edges according
+        // to that function, even if this is a multigraph. If unspecified,
+        // defer parallel edge handling to `add_edge_no_cycle_check`.
+        if let Some(merge_fn) = weight_combo_fn {
+            macro_rules! merge_parallel_edges {
+                ($edges:ident) => {{
+                    let mut node_to_weight = HashMap::new();
+                    for (node, weight) in $edges {
+                        match node_to_weight.remove(&node) {
+                            Some(existing_weight) => {
+                                let combined_weight = merge_fn
+                                    .call1(py, (weight, existing_weight))?;
+                                node_to_weight.insert(node, combined_weight);
+                            }
+                            None => {
+                                node_to_weight.insert(node, weight);
+                            }
+                        }
```

Comment:
I think the following should be slightly faster without the need to first remove and then re-insert a node. I see in previous commits you did something similar with `and_modify/or_insert`, was there a reason for a change?

```suggestion
                        match node_to_weight.entry(node) {
                            Occupied(entry) => {
                                *entry.into_mut() = merge_fn
                                    .call1(py, (weight, entry.get()))?;
                            }
                            Vacant(entry) => {
                                entry.insert(weight);
                            }
                        }
```

Reply:
Ah yeah, something like this could be nice! I got rid of the `and_modify` stuff because it seemed like I'd then need to store the `PyObject` as a `PyResult<PyObject>` in the map, since we couldn't use the `?` inside the closure. Your suggestion doesn't have that problem :). I used `remove` because it moves the value out of the map, which meant I could avoid `.clone_ref(py)` on the value, but I _think_ what you have should work too since you've got a non-mutable ref of `entry`, it seems.

Reply:
Done in 3278664.